### PR TITLE
monitoring: add query validation with template variables

### PR DESCRIPTION
**Pull request by bobheadxi**
Introduces the PromQL parser to the monitoring generator, and some adjustments to strip queries of template variables (`$variable`) to allow the parser to parse queries. This allows us to:

- **access to the query abstract syntax tree to enable universal query manipulations: https://github.com/sourcegraph/sourcegraph/pull/41644**
- validate that referenced variables are actually defined on a dashboard (unsubstituted variables will cause the parser to throw an error)
- validate that all queries are valid Prometheus queries
- in-code documentation of what query option variables look like (because a sample value is now required)

This is safe for 4.0 because it does not change anything in the output, just introduces additional validation.

## Test plan

`go generate ./monitoring` passes

### CODE REVIEW

**Comment by bobheadxi** [comment]
Current dependencies on/for this PR:
* main
  * **PR #41642** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/41642" target="_blank">Graphite</a> 
    * **PR #41643** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/41643" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/41643?utm_source=stack-comment).

**Review by jhchabran — COMMENTED** [review]
Did a first pass, it's a bit abstract at first without being familiar with that part of the codebase.

@bobheadxi Do we have automated tests for this? I saw you ran some manual tests, but as this is growing, it sounds quite concerning to not have those.

**Comment by bobheadxi** [comment]
> Do we have automated tests for this? I saw you ran some manual tests, but as this is growing, it sounds quite concerning to not have those.

The default generate task and Grafana image build is effectively a test of the core path - I'll add something specific for just the new `promql` package

**Comment by bobheadxi** [comment]
Also adding a super simple e2e generator test further up the stack: https://github.com/sourcegraph/sourcegraph/pull/41766

**Comment by jhchabran** [comment]
> Also adding a super simple e2e generator test further up the stack: #41766

Awesome! Btw, my previous comment was a bit expedited, there's a lot of work in there and what's enabled by it is pretty awesome 😊 I'm just scratching my head a bit on the complexity of this but I haven't thought out what else we could do so I didn't want to bring that in :)

File: monitoring/monitoring/variables.go [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+}
+
+type variableApplier struct{ vars []ContainerVariable }
+
+var _ promql.VariableApplier = &variableApplier{}
+
+func newVariableApplier(vars []ContainerVariable) promql.VariableApplier {
+	return &variableApplier{
+		vars: append(vars,
+			// Make sure default Grafana variables are applied too
```
Comment (jhchabran) [inline monitoring/monitoring/variables.go:207]: more docstring for context ?

File: monitoring/monitoring/variables.go [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+				},
+			}),
+	}
+}
+
+func (a *variableApplier) ApplyDefaults(expression string) string {
+	for _, v := range a.vars {
+		varKey := "$" + v.Name
+
+		// If the expression uses the variable in a quoted context ("$var") then it's=
```
Comment (jhchabran) [inline monitoring/monitoring/variables.go:221]: typo :D

File: monitoring/monitoring/variables_test.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestVariableExampleValue(t *testing.T) {
+	// Numbers get replaced
+	assert.Equal(t, "1234m",
```
Comment (jhchabran) [inline monitoring/monitoring/variables_test.go:11]: nit: comment that 1234m is the default placeholder

**Review by jhchabran — APPROVED** [review]
Reviewed while pairing, small nits, but LGTM !

**Pull request merged**